Fix AddUpdateAPIService for openapiv2

Jefftree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:

A couple of issues were discovered and they've been combined into this PR since they all need a backport.

For https://github.com/kubernetes/kubernetes/issues/120758: Triggered by a race when a GET to /openapi/v2 is called between https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L227 and https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L230

Fix is to initialize the cache with empty data so aggregation won't have problems.

For https://github.com/kubernetes/kubernetes/issues/120739: AddUpdateAPIService was adding and also updating the apiservice spec. The apiservice spec update should be done async by the controller.

Fix is to not update the spec in the addupdateapiservice flow during startup. This is handled async by the openapi controller.

I've discovered https://github.com/kubernetes/kubernetes/issues/120878 while working on the above two issues: The code path when an apiservice exists and needs to be updated was not handled properly. I've refactored the CacheableDownloader to have an new function `UpdateHandler()` and call that in `AddUpdateAPIService()` if the apiservice already exists. I've also removed the `decorateError` function and integrated the better logging directly in the downloader.

I've fixed the `AddUpdateAPIService` flow and added integration tests to simulate the cases described in #120758 and #120739. Added additional unit tests to catch #120878. Tried to keep the changes self contained since we need to backport.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/120758
Fixes https://github.com/kubernetes/kubernetes/issues/120739
Fixes https://github.com/kubernetes/kubernetes/issues/120878

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix regression with adding aggregated apiservices panicking and affected health check introduced in release v1.28.0
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jiahuif commented:
/assign @apelisse

dims commented:
@Jefftree will use this PR to try some CI jobs. hope that's ok :)

dims commented:
/test pull-kubernetes-e2e-containerd-gce
/test pull-kubernetes-node-e2e-containerd

liggitt commented:
Please add the exact patch version where the regression was introduced into the release note

Jefftree committed:
Fix 120758 - prevent cache Load on uninitialized spec

Jefftree committed:
Fix 120739 - Prevent AddUpdateAPIService from blocking on updating spec

Jefftree committed:
Fix 120878 - Refactor cacheabledownloader so handler can be updated and fix AddUpdateAPIService to update handler

Jefftree committed:
Unit & integration tests for OpenAPIV2 AddUpdateAPIService

Jefftree commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -224,6 +224,7 @@ func (s *specAggregator) AddUpdateAPIService(apiService *v1.APIService, handler
 			apiService: *apiService,
 			downloader: decorateError(apiService.Name, NewCacheableDownloader(s.downloader, handler)),
 		}
+		s.specByAPIServiceName[apiService.Name].spec.Store(cached.Result[*spec.Swagger]{Err: fmt.Errorf("spec for apiservice %s is not yet available", apiService.Name)})
```
We only serve things we know. Errors are logged and skipped: https://github.com/kubernetes/kubernetes/blob/bc10564de4d5bfc05049f750805c1680fba0aeb1/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L161-L166

Jefftree commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -21,31 +21,54 @@ import (
 	"k8s.io/kube-openapi/pkg/validation/spec"
 )
 
+type CacheableDownloader interface {
```
How would we update a CacheableDownloader with a new handler on the update case? We could call Get() on the old downloader to get the last spec and etag, but that would be blocking.

apelisse replied:
I guess my point is that the cacheableDownloader is supposed to cache "on-demand", but nothing about this is on-demand, is it? it's done periodically.

apelisse replied:
yeah, right, thanks

Jefftree commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -218,14 +217,22 @@ func (s *specAggregator) AddUpdateAPIService(apiService *v1.APIService, handler
+		if existingSpec.downloader != nil {
+			existingSpec.downloader.UpdateHandler(handler)
+		} else {
+			klog.Warningf("Should not happen: downloader for apiservice %s is nil", apiService.Name)
```
ok, I'll update this to directly access the downloader.

apelisse replied:
Whichever, a panic would be ideal IMO

apelisse replied:
Please address this one and I'll lgtm. Thanks

Jefftree replied:
Fixed.

Jefftree commented on test/integration/apiserver/openapi/openapi_apiservice_test.go:
```diff
@@ -0,0 +1,239 @@
+		// Force a 5 second APIService openapi response time.
+		// This cannot be longer because goleak checker will
+		// terminate long running goroutines stuck in sleep.
+		time.Sleep(time.Second * 5)
```
:(

```
E0925 14:39:23.841987   20047 etcd.go:221] "EtcdMain goroutine check" err=<
	found unexpected goroutines:
	[Goroutine 3964 in state sleep, with time.Sleep on top of the stack:
	goroutine 3964 [sleep]:
	time.Sleep(0x12a05f200)
		/usr/local/go/src/runtime/time.go:195 +0x118
	k8s.io/kubernetes/test/integration/apiserver/openapi.TestSlowAPIServiceOpenAPIDoesNotBlockHealthCheck.func1({0x10779ab40, 0x14008954460}, 0x106dd06e0?)
		/Users/jying/workspace/kubernetes/test/integration/apiserver/openapi/openapi_apiservice_test.go:56 +0x3c
```

apelisse commented:
`AddUpdate` and `Update` names are a little weird now. Shouldn't we remove`AddUpdate` into `Add`?

Jefftree commented:
> `AddUpdate` and `Update` names are a little weird now. Shouldn't we remove`AddUpdate` into `Add`?

AddUpdate is correct, it either adds or updates the apiservice object (not the spec).

Jefftree commented:
I've updated the test to avoid using the sleep and just block until `ctx.Done()` which is equivalent to never returning before the test ends. Also updated https://github.com/kubernetes/kubernetes/pull/120814/files#diff-5aea25bc7e6557f173ac9d38ed1dca53f412d908a2f32f71720ef3591ab4a95fR142-R144 to skip updates for local apiservices (copied from AddUpdateAPIService https://github.com/kubernetes/kubernetes/blob/6427c4d3a9a5139325f986ce0ea72404a466caae/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L214-L216)

Jefftree commented:
/retest

```
{ failed [FAILED] PersistentVolume hostpath-plc2p not in phase Available within 30s
In [BeforeEach] at: test/e2e/storage/pv_protection.go:87 @ 09/27/23 16:46:12.385
}
```

apelisse commented:
/lgtm
/approve

Jefftree commented:
/cc @sttts @liggitt

liggitt commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:
```diff
@@ -139,7 +139,10 @@ func (c *AggregationController) AddAPIService(handler http.Handler, apiService *
 // UpdateAPIService updates API Service's info and handler.
 func (c *AggregationController) UpdateAPIService(handler http.Handler, apiService *v1.APIService) {
-	if err := c.openAPIAggregationManager.AddUpdateAPIService(apiService, handler); err != nil {
+	if apiService.Spec.Service == nil {
```
this doesn't handle aggregating openapi for custom resources?

apelisse replied:
IIUC, the update here is just to refresh aggregated apiservers by downloading every once in a while. CRDs OpenAPI is refreshed on demand.

Jefftree replied:
Yep exactly as @apelisse mentioned. This nil check was part of AddUpdateAPIService, but since we don't call that here anymore, add the check here too.

liggitt commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -21,31 +21,54 @@ import (
+	d.handler = handler
+}
 
 func (d *cacheableDownloader) Get() (*spec.Swagger, string, error) {
```
Get is long-running, right? the mutex addition makes that block UpdateHandler, which now blocks specAggregator#AddUpdateAPIService

That means a hang fetching one service's openapi can block processing of other services, right?

Jefftree replied:
hmm that's true. What if we have the mutex only be required when we fetch the handler instead of the entire get operation?

liggitt replied:
is the only thing we need to guard the handler, or is the etag also concurrently modifiable? (can Get be called from multiple threads?)

Jefftree replied:
We currently don't call Get() from multiple threads but the defacto `cached.Value` interface that we follow is supposed to be thread safe (https://github.com/kubernetes/kube-openapi/blob/master/pkg/cached/cache.go#L17-L22). 

@apelisse I think this ties into what you mentioned earlier that since we don't evaluate aggregated apiservices lazily, maybe we shouldn't follow cached library interface?

liggitt replied:
leaving this open before merge just to get @apelisse's feedback

apelisse replied:
If you don't then you'll have to deal with the etag manually, I'm not sure it would better.

liggitt commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -218,16 +217,21 @@ func (s *specAggregator) AddUpdateAPIService(apiService *v1.APIService, handler
 		}
+		specInfo.spec.Store(cached.Result[*spec.Swagger]{Err: fmt.Errorf("spec for apiservice %s is not yet available", apiService.Name)})
+		s.specsByAPIServiceName[apiService.Name] = specInfo
 		s.openAPIVersionedService.UpdateSpecLazy(s.buildMergeSpecLocked())
```
I guess buildMergeSpecLocked already called Get(), which meant a slow request / timeout of one service would keep the specAggregator#UpdateAPIServiceSpec loop from processing other API services. Is that new in 1.28 or was it existing?

Jefftree replied:
specAggregator#UpdateAPIServiceSpec blocking for each apiservice is pre-existing. The controller is single threaded so the UpdateAPIServiceSpec loop processes one apiservice at a time in a blocking manner. builtMergeSpecLocked calling Get() is a new addition, it is only called for local apiservices and from observation local apiservices are very unlikely to block and cause problems.

liggitt replied:
> specAggregator#UpdateAPIServiceSpec blocking for each apiservice is pre-existing

does that mean this PR does not resolve https://github.com/kubernetes/kubernetes/issues/120739#issuecomment-1725875180

> The issue occurs only after v1.28 may be because https://github.com/kubernetes/kubernetes/pull/118212 changed c.openAPIAggregationManager.AddUpdateAPIService to download the latest spec synchronously?

Jefftree replied:
when an APIService is added, only AddUpdateAPIService is called. There was a bug where AddUpdateAPIService called UpdateAPIServiceSpec that was fixed by this commit targeting #120739: https://github.com/kubernetes/kubernetes/pull/120814/commits/61603d36d738dddf3a7f244e0f8630f9ebeeffaf.

UpdateAPIServiceSpec is called by the openapi controller and is not in the path for readiness, and we are essentially leaving the behavior of UpdateAPIServiceSpec unchanged from before the regression.

liggitt replied:
sorry, I think I copied the wrong bit from your comment

> builtMergeSpecLocked calling Get() is a new addition

Aren't we still doing that in the current state of this PR?

> it is only called for local apiservices and from observation local apiservices are very unlikely to block and cause problems.

Where is it limited to local apiservices? This function appears to only be handling *remote* API services (it short-circuits if apiService.Spec.Service == nil)

Jefftree replied:
> Aren't we still doing that in the current state of this PR?

Yes, `buildMergeSpecLocked` calls `Get()` on local apiservices. The problem described in the issue is for remote apiservices which leads to your next comment

> Where is it limited to local apiservices? This function appears to only be handling remote API services (it short-circuits if apiService.Spec.Service == nil)

`buildMergeSpecLocked()` calling `Get()` is limited to local apiservices. The `UpdateAPIServiceSpec()` function only affects remote apiservices (as seen with the short circuit). `UpdateAPIServiceSpec()` is not called within the critical path of readiness (only AddUpdateAPIService).

For readiness: 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L330-L337
The apiservice controller syncs all existing apiservices before closing the ready channel and calls AddAPIService for each existing apiservice:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go#L112

which goes through
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L478

and then reaches openapi's AddUpdateAPIService.

liggitt replied:
> `buildMergeSpecLocked()` calling `Get()` is limited to local apiservices.

It took me a while to understand why that is true...
1. we populate a cache entry for this remote service here containing an error that the spec wasn't synced yet
```go
		specInfo.spec.Store(cached.Result[*spec.Swagger]{Err: fmt.Errorf("spec for apiservice %s is not yet available", apiService.Name)})
		s.specsByAPIServiceName[apiService.Name] = specInfo
```
2. buildMergeSpecLocked accumulates cached specs:
```
	for i, apiService := range apiServices {
		caches[i] = &(s.specsByAPIServiceName[apiService.Name].spec)
	}
```
3. buildMergeSpecLocked merges calling .Get on each one:
```go
	return cached.MergeList(func(results []cached.Result[*spec.Swagger]) (*spec.Swagger, string, error) {
		var merged *spec.Swagger
		etags := make([]string, 0, len(results))
		for _, specInfo := range results {
			result, etag, err := specInfo.Get()
```
4. the result we just included will just return an error:
```go
func (r Result[T]) Get() (T, string, error) {
	return r.Value, r.Etag, r.Err
}
```
5. which buildMergeSpecLocked will log and continue:
```go
			if err != nil {
				// APIService name and err message will be included in
				// the error message as part of decorateError
				klog.Warning(err)
				continue
			}
```

I didn't realize we were adding two different implementations of Get() into `specsByAPIServiceName`, and that the one for remote services had a non-blocking Get() implementation

apelisse replied:
you folks are confusing to read, `buildMergeSpecLocked` almost does nothing, the closure that it builds does most of the work ... asynchronously. But yes, otherwise

> I didn't realize we were adding two different implementations of Get() into specsByAPIServiceName, and that the one for remote services had a non-blocking Get() implementation

That's correct.

liggitt commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -21,31 +21,54 @@ import (
 func (d *cacheableDownloader) Get() (*spec.Swagger, string, error) {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	spec, etag, err := d.getLocked()
```
it's a little confusing to call `getLocked` when we're only under an RLock ... it's not obvious what mutex is guarding, but we assign to etag and spec inside `getLocked`

apelisse replied:
I agree that's wrong. I don't know if it'll trigger with the current way things are implemented, but that's still a problem with that object.

Jefftree replied:
Yeah I agree. I was originally thinking of only protecting the handler with this lock, but I can see how this can be interpreted as etag/spec as well.

Jefftree replied:
I've changed the mutex to just a regular mutex since get is only called from one thread and rlock has no performance benefit. Also modified the lock to only be active when fetching the handler rather than the entire get operation.

liggitt commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -218,16 +219,21 @@ func (s *specAggregator) AddUpdateAPIService(apiService *v1.APIService, handler
 	s.mutex.Lock()
```
if an API Service changes from remote to local (service != nil → service == nil), will the short-circuit return above prevent cleanup from happening? should we call RemoveAPIService in that case?

just noticed this, it's pre-existing in 1.27, so probably best to handle separately in a follow up

Jefftree replied:
ack, will do that in a follow up

liggitt replied:
thanks, link the issue here when you open it to close the loop

Jefftree commented:
/retest

liggitt commented:
lgtm but will defer to @apelisse on https://github.com/kubernetes/kubernetes/pull/120814#discussion_r1347591395 before merge

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -21,32 +21,62 @@ import (
+	return spec, etag, err
+}
+
+func (d *cacheableDownloader) getWithHandler(h http.Handler) (*spec.Swagger, string, error) {
```
is there a `get` without handler? The name sounds like there would be. But wait, I think I see where this is coming from. You shouldn't have to receive the handler here.

Jefftree replied:
I've updated with the atomic pointer, but kept this function. Separating the two functions lets us decorate the error with the apiservice name. lmk if you have thoughts, thanks!

apelisse replied:
Sure, I still think it's weird to pass to `getWithHandler` one of it's own member though.

apelisse replied:
Maybe just do ...
```golang
spec, etag, err := d.get()
...
func (d *cacheableDownloader) get() (*spec.Swagger, string, error) {
    h := *d.handler.Load()
...
```

Jefftree replied:
Updated

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -21,32 +21,62 @@ import (
 	etag       string
 	spec       *spec.Swagger
+
+	// mutex protects the handler which can be replaced
+	mutex sync.Mutex
```
if the mutex protects only one variable, why don't you use an Atomic pointer instead to protect the handler?

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -21,32 +21,62 @@ import (
+	d.handler = handler
+}
+
+func (d *cacheableDownloader) getHandler() http.Handler {
```
If you use an atomic handler, then I think this is mostly unneeded, which makes the whole function below simpler (i.e. getting the handler is nothing more than `handler.Load()`)

apelisse reviewed (commented):
My comments are nits and don't have to be fixed here since they mostly relate to style and have no behavior impact. LGTM for me, thanks.

liggitt commented:
/retest
/approve

liggitt commented:
/hold in case you want to incorporate @apelisse's comments here before backport

apelisse commented:
Thanks
/lgtm

liggitt commented:
/hold cancel
